### ISSUE TITLE
 tfp.distributions.GaussianProcess Optimization

### ISSUE THREAD

**Issue by easysoft2k15** [tensorflow/probability#450]
Hi,
I'm playing with tfp.distributions.GaussianProcess in Tensowflow 2.0.
This is the simple model I came up with:

`#Toys data`
`#------------------------------------------------------------------------------------`
`x=tf.linspace(0.,10.,100)`
`y=2.*x+tfp.distributions.Normal(0.,1.).sample(100)`
`x_=tf.reshape(x,[-1,1])`

`#Model definition`
`#------------------------------------------------------------------------------------`
`amplitude=tf.Variable(10.,name="amplitude")`
`length_scale=tf.Variable(0.5,name="length_scale")`
`k=tfp.positive_semidefinite_kernels.MaternThreeHalves(amplitude=amplitude,length_scale=length_scale)`
`gp=tfp.distributions.GaussianProcess(kernel=k,index_points=x_)`

`#Optimization`
`#------------------------------------------------------------------------------------`
`loss=lambda: -gp.log_prob(y)`
`opt=tf.optimizers.Adam(learning_rate=0.1)`
`losses=[]`
`for i in range(1001):`
    `opt.minimize(loss,var_list=[amplitude,length_scale])`
   ` if i % 50 == 0:`
        `losses.append(loss())`

The error I get is:

**ValueError: No gradients provided for any variable: ['amplitude:0', 'length_scale:0'].**

From my understanding this is due to Eager Execution. Somehow the two *tf.Variable* (**amplitude** and **length_scale**) are *"resolved"* after model definition and the loss function is non depending on them anymore (they are kind on baked into the model at creation-time).
in other words the parameters **gp.kernel.amplitude** and **gp.kernel.length_scale** seems not to be *"trainable"*.
What is the recommended way to solve this?
Thank You

**TF and TFP version:**
**tf-nightly-2.0-preview           2.0.0.dev20190605**
**tfp-nightly                             0.8.0.dev20190606**

**Comment by brianwa84** [comment]
https://github.com/tensorflow/probability/commit/bfe97723f0bc040491c05dc68348b4be5671fed3
which might have made it into today's nightly should be very relevant here.

**Comment by brianwa84** [comment]
The other alternative for your situation is to put a `def loss():` right after defining the variables and before `k = ...`

**Comment by easysoft2k15** [comment]
I tried this:

`amplitude=tf.Variable(10.,name='amplitude')`
`length_scale=tf.Variable(0.5,name='length_scale')`

`def loss():`
`   return -gp.log_prob(y)`

`k=tfp.positive_semidefinite_kernels.MaternThreeHalves(amplitude=amplitude,length_scale=length_scale)`
`gp=tfp.distributions.GaussianProcess(kernel=k,index_points=x_)`

but I get the same error.

I see what You are doing with **trainable_variables** and this seems definitely the right thing to do.
At the moment though *gp.trainable_variables* seems to be a empty tuple probably because the GaussianProcess distribution has not been modified yet.
I'll wait for the change (i'm only playing with it at the moment)
Thank You for pointing me into the right direction.

**Comment by brianwa84** [comment]
The k= and gp= lines need to be inside `def loss()`.

Brian Patton | Software Engineer | bjp@google.com

**Comment by easysoft2k15** [comment]
I see what You meant. Which basically means that I'm, recreating the GaussianProcess each time I do a gradient step. I just wander what is the overhead for the recreation.
In any case It works!
Thank You

**Comment by brianwa84** [comment]
There is certainly overhead, especially insofar as creating the GP likely
involves a cholesky decomposition (N^3 in your data size). However, if you
want gradients to your variables and the variables are influencing the
kernel that drives the cov matrix, then you will have to pay that N^3 at
each gradient step regardless.

Brian Patton | Software Engineer | bjp@google.com

**Comment by csuter** [comment]
Also: If you decorate the loss with @tf.function you won't pay the python
object creation cost on each gradient step anymore.
